File: P2/PartitionStream.cpp
// Author: Ayimen H.
// Date: Oct 12, 2024
// Revision: 0

// Class Invariant:
// Array is zero indexed and client should be aware of this when passing in partition key

#include "PartitionStream.h"

using namespace std;

PartitionStream::PartitionStream(MsgStream*& driverPartitionStream, unsigned driverNumPartitions)
{
    if (driverPartitionStream == nullptr || driverNumPartitions == 0)
    {
        throw invalid_argument("Unacceptable partition stream provided!");
    }

    partitionStream = driverPartitionStream;
    driverPartitionStream = nullptr;

    this -> numPartitions = driverNumPartitions;
}

void PartitionStream::writeMessage(int partitionKey, const string& message)
{
    if (partitionKey > numPartitions)
    {
        throw out_of_range("Invalid Partition Key!");
    }
    partitionStream[partitionKey].AppendMessage(message);
}

string* PartitionStream::readMessage(int partitionKey, int startingMessage, int endingMessage)
{
    if (partitionKey > numPartitions)
    {
        throw out_of_range("Invalid Partition Key!");
    }
    return partitionStream[partitionKey].ReadMessages(startingMessage, endingMessage);
}

int PartitionStream::QueryPartitionKey()
{
    int QueriedPartitionKey = numPartitions;
    return QueriedPartitionKey;
}

PartitionStream::PartitionStream(const PartitionStream& src)
{
    numPartitions = src.numPartitions;
    partitionStream = new MsgStream[numPartitions];

    for (unsigned i = 0; i < numPartitions; i++)
    {
        partitionStream[i] = src.partitionStream[i];
    }
}

PartitionStream& PartitionStream::operator= (const PartitionStream& src)
{
    if (this == &src)
    {
        return *this;
    }

    delete[] partitionStream;
    numPartitions = src.numPartitions;

    partitionStream = new MsgStream[numPartitions];
    for (unsigned i = 0; i < numPartitions; i++)
    {
        partitionStream[i] = src.partitionStream[i];
    }

    return *this;
}

PartitionStream::PartitionStream(PartitionStream&& src) noexcept
{
    numPartitions = src.numPartitions;
    partitionStream = src.partitionStream;

    src.numPartitions = 0;
    src.partitionStream = nullptr;
}

PartitionStream& PartitionStream::operator= (PartitionStream&& src) noexcept
{
    if (this == &src)
    {
        return *this;
    }

    swap(numPartitions, src.numPartitions);
    swap(partitionStream, src.partitionStream);

    return *this;
}

void PartitionStream::reset()
{
    for (unsigned i = 0; i < numPartitions; i++)
    {
        partitionStream[i].Reset();
        delete &partitionStream[i];
    }
    numPartitions = 0;
}

PartitionStream::~PartitionStream()
{
    delete[] partitionStream;
}

// Implementation Invariant:
// partitionStream must be valid after initialization
// numPartitions must reflect the actual number of MsgStream objects

File: P5/P5/PartitionStream.cs
// Author: Ayimen H.
// Date: Nov 25, 2024
// Revision: 0th Revision


// Class Invariant:
// basePartitionStream holds the initial copies of partitionStream for use in reset operations
// ^ for numPartitions
// Array is zero indexed and client should be aware of this when passing in partition key

using System;
using System.IO;
using System.Runtime.CompilerServices;

namespace P5;

public class PartitionStream: IDisposable
{ 
    private MsgStream[] partitionStream;
    private int numPartitions;
    
    private MsgStream[] basePartitionStream;
    private int baseNumPartitions;

    private bool isDisposed = false;
    
    // PostCondition: Initializes PartitionStream with deep copies of partition streams
    public PartitionStream(MsgStream[] driverPartitionStream, int driverNumPartitions)
    {
        if (driverPartitionStream == null || driverNumPartitions == 0)
        {
            throw new ArgumentException("Unacceptable partition stream provided!");
        }

        partitionStream = driverPartitionStream;
        driverPartitionStream = null;
        numPartitions = driverNumPartitions;
        
        basePartitionStream = new MsgStream[numPartitions];
        baseNumPartitions = numPartitions;

        for (int i = 0; i < numPartitions; i++)
        {
            basePartitionStream[i] = partitionStream[i].DeepCopy();
        }
    }
    
    public void WriteMessage(int partitionKey, string message)
    {
        try
        {
            if (partitionKey > numPartitions || partitionKey < 0)
            {
                throw new ArgumentOutOfRangeException("Invalid Partition Key!");
            }
            partitionStream[partitionKey].AppendMessage(message);
        }
        
        catch (Exception e)
        {
            Console.WriteLine($"Error in WriteMessage: {e.Message}");
            throw;
        }
    }
    
    public string[] ReadMessage(int partitionKey, int startingMessage, int endingMessage)
    {
        try
        {
            if (partitionKey > numPartitions || partitionKey < 0)
            {
                throw new ArgumentOutOfRangeException("Invalid Partition Key!");
            }
            return partitionStream[partitionKey].ReadMessages(startingMessage, endingMessage);
        }
        
        catch (Exception e)
        {
            Console.WriteLine($"Error in ReadMessage: {e.Message}");
            throw;
        }
    }
    
    public PartitionStream DeepCopy()
    {
        MsgStream[] copiedStreams = new MsgStream[numPartitions];
        for (int i = 0; i < numPartitions; i++)
        {
            copiedStreams[i] = partitionStream[i].DeepCopy();
        }
    
        PartitionStream copy = new PartitionStream(copiedStreams, numPartitions);
        return copy;
    }
    
    public int QueryPartitionKey()
    {
        int QueriedPartitionKey = numPartitions;
        return QueriedPartitionKey;
    }
    
    // PostCondition: Restores partitionStream to its initial base state
    public void Reset()
    {
        for (int i = 0; i < numPartitions; i++)
        {
            basePartitionStream[i].Reset();
            partitionStream[i] = basePartitionStream[i];
        }
        
        numPartitions = baseNumPartitions;
    }
    
    public void Dispose()
    {
        Dispose(true);
        GC.SuppressFinalize(this);
    }

    protected virtual void Dispose(bool isDisposing)
    {
        if (isDisposing)
        {
            if (isDisposing)
            {
                if (partitionStream != null)
                {
                    for (int i = 0; i < partitionStream.Length; i++)
                    {
                        MsgStream msgStream = partitionStream[i];
                        if (msgStream is IDisposable disposableStream)
                        {
                            disposableStream.Dispose();
                        }
                    }
                }

                if (basePartitionStream != null)
                {
                    for (int i = 0; i < basePartitionStream.Length; i++)
                    {
                        MsgStream msgStream = basePartitionStream[i];
                        if (msgStream is IDisposable disposableStream)
                        {
                            disposableStream.Dispose();
                        }
                    }
                }
            }
            
            isDisposed = true;
        }
    }
}

// Implementation Invariant:
// Append and Read operations are restricted to valid partition indices